### PR TITLE
Additional swagger features

### LINKED ISSUE #2504: Additional feature for new Swagger UI
Swagger v3 has not only new UI but new realization as well. The main idea is to know can we realize this features and do that. 

### Additional features
- API call with api_key via Swagger UI (Related #2497)
- Try-out method
- Calling request via Proxy Backend
- Rate limiting info in headers (Related #2498). This feature is not available in old Swagger but can we implement in new one?

### LINKED ISSUE #2548: API Profile Documentation: Page needs to be reloaded to close Authorize dialogue
# Reproduction Steps:
* Go to nightly.apinf.io
* Go to documentation tab of an API that has a valid swagger (Try APInf User API)
* click on the Authorize button

# Result
* Authorize dialog doesn't have any close button to close the dialog.
* Page needs to be reloaded or navigated back to the previous page to close the dialog
doctabauthorization

# environment
* google chrome
* windows 8

### PR DESCRIPTION
**Pull request by marla-singer**
Closes #2504 
Closes #2548
## Changes
- Allows API manager to set supported Submit methods for call requests
- If API is connected to proxy then replace host and basePath to Proxy URL and Proxy Backend path
- If API key is required for requests then create a security definitions for Proxy Authorization
- If a user has an API key then auto-complete value into Authorization form
- Displays SVG icons

## For reviewer
Additional features for Swagger UI v3

**1. Control Try out button**
An API manager can limit to users call specified methods using for that Manage module window

## Steps
1. Open managed API -> Documentation tab
2. Click on Manage button
3. Upload Swagger file
4. Check needed methods below label "Allow try-out for following methods:"

### Expected
- if a method is checked then a user can Try out this method (Try out button is shown)
joxi_screenshot_1495801034308 1
- if all methods are checked then all methods are allowed for users
- if no one methods are checked then no methods are allowed
joxi_screenshot_1495801034308

**2. Redirect request to Proxy if API is connected to**

### Cases

**1. If API is connected to proxy**

### Expected
- all request from Swagger must call via Proxy

joxi_screenshot_1495800717915

**1. If API is not connected to proxy**

### Expected
- all request from Swagger must call to an original API

joxi_screenshot_1495800717915 1

**3. Autocomplete API key for Proxy Authorization**

API-umbrella proxy has own authorization via query parameter and via header.

For this reason, if API is connected to proxy then there needs double authentication: one for proxy and another one for original API

### Cases
**1. API is connected to proxy and API key is disabled for requests**

### Steps
1. Open managed API  -> Proxy tab
2. Fill up data
3. Check Disabled API key as true

### Expected
- if a user has (or not) API key anyway Additional Auth form doesn't exist

**2. API is connected to proxy and API key is required for requests**
   
1. Open managed API  -> Proxy tab
2. Fill up data
3. Check Disabled API key as false

### Expected
- Additional Auth form exists
- if a user has API key then the value is auto-completed
- if a user has not API key then the value is empty

joxi_screenshot_1495800542543 1

### CODE REVIEW

**Comment by mauriciovieira** [comment]
@marla-singer where are you getting the minified swagger bundle files from?

File: api_docs/client/swagger_ui/content/content.js [hunk 1]
```diff
@@ -1,30 +1,139 @@
...
+  const proxyBackend = ProxyBackends.findOne({ apiId: api._id });
+
+  // Placeholders
+  let proxyBasePath;
+  let proxyHost;
+  let apiKeyValue;
+  let apiKeyDisabled;
+
+  // Make sure API has proxy backend configuration
+  if (proxyBackend) {
```
Comment (mauriciovieira) [inline api_docs/client/swagger_ui/content/content.js]: I don't have any proxyBackend in my installation. What should happen? Just go silent?

File: api_docs/client/swagger_ui/svg_icons.html [hunk 2]
```diff
@@ -0,0 +1,40 @@
+<!-- Copyright 2017 Apinf Oy
+This file is covered by the EUPL license.
+You may obtain a copy of the licence at
+https://joinup.ec.europa.eu/community/eupl/og_page/european-union-public-licence-eupl-v11 -->
+
+<template name="swaggerUiSvgIcons">
```
Comment (mauriciovieira) [inline api_docs/client/swagger_ui/svg_icons.html]: This part is 👍 I could verify locally that we have the icons now.

File: api_docs/client/swagger_ui/content/content.js [hunk 3]
```diff
@@ -1,30 +1,139 @@
...
+    // Get URL of related proxy
+    proxyHost = proxyBackend.proxyUrl();
+
+    // Make sure Proxy URL exists
+    if (proxyHost) {
```
Comment (mauriciovieira) [inline api_docs/client/swagger_ui/content/content.js]: What should happen case there is no proxyHost?

**Comment by mauriciovieira** [comment]
out

**Comment by marla-singer** [comment]
> where are you getting the minified swagger bundle files from?

@mauriciovieira I forked Swagger repo and build a new version of `budle.js` and move it to apinf project. I don't really like this stuff and I think maybe we should storage this file to CDN storage or something like that.

I will prepare a paper about Swagger to describe what is expected while working with Swagger and will give you a link to the paper

**Comment by marla-singer** [comment]
@mauriciovieira Added helpful information in Description box

**Comment by mauriciovieira** [comment]
@marla-singer ok. I will test it today. If it's okay, I'll merge.

**Comment by mauriciovieira** [comment]
Well, I am still doing something wrong or I don't know how to setup the proxy locally. (Well, as I said a few times, we should have a simple setup command from scratch, something like `yarn seeds` to seed basic information for any developer).

What I did was 

* meteor reset
* meteor run on this branch
* got the api-umbrella settings from nightly.apinf.io (API Key and API Token), and registered locally

screen shot 2017-05-29 at 18 05 57

* chose an [API](https://nightly.apinf.io/apis/kasviplanktonrajapinta) from apinf.io.
* read documentation, [test locally in command line using curl](https://gist.github.com/mauriciovieira/d7bac067746afc1bd83a5ae6461e554e#file-curl_command_and_output)
* created a [swagger file](https://gist.github.com/mauriciovieira/d7bac067746afc1bd83a5ae6461e554e) for it using http://editor2.swagger.io/
* Add api
* Api URL http://api.digitransit.fi/geocoding/v1/
* Documentation:

screen shot 2017-05-29 at 18 13 18

* Proxy - I think it saved to api-umbrella correctly but I don't know how to check that:

screen shot 2017-05-29 at 18 16 19

* Execution in swagger ui

screen shot 2017-05-29 at 18 17 28

Is there anything that I did wrong?

Can @brylie help by also validating this ui?

**Comment by mauriciovieira** [comment]
Oh, I see what I did wrong. After I saved the proxy, I got

screen shot 2017-05-29 at 18 21 28

But 

```$ curl "https://nightly.apinf.io:3002/digitransit_geocoding/"                                                                                     
{
  "error": {
    "code": "API_KEY_MISSING",
    "message": "No api_key was supplied. Get one at https://nightly.apinf.io:3002"
  }
}```

means that something is wrong in API Umbrella integration, because I did check ```Disable API key requirement``` and saved, but I got the same error on the command line after that.

![screen shot 2017-05-29 at 18 25 19](https://cloud.githubusercontent.com/assets/95258/26562380/330cce96-449c-11e7-8cd2-c647d404ef65.png)

The ideal solution here would be not depending on a remote api umbrella instalation, but locally (docker) (just opened #2612).

Again, if any other @apinf/developers  can review this ticket faster than me, please do it.

**Comment by philippeluickx** [comment]
Good work there, looking forward to having this on production!

Concerning the bundle.js, it's indeed not the best workflow. Is CDN best option? There is one for swagger-ui?
https://cdnjs.com/libraries/swagger-ui

Otherwise my preference would go to get this in through npm? https://www.npmjs.com/package/swagger
Would that also work?

This can be a seperate ticket if work needs to be done.

**Comment by brylie** [comment]
> Concerning the bundle.js, it's indeed not the best workflow. Is CDN best option? There is one for swagger-ui?
> https://cdnjs.com/libraries/swagger-ui

We keep the swagger files in our server static files (`/public`) directory, which is not included in the main application bundle:

> All files inside a top-level directory called public/ are served as-is to the client.
>
> Meteor Docs - Application Structure - [Special Directories](https://guide.meteor.com/structure.html#special-directories)

What would be some pros and cons of using a CDN for Swagger UI? How would we include the CDN file reference on the API Profile view, to guarantee the documentation browser would render?

> Otherwise my preference would go to get this in through npm? https://www.npmjs.com/package/swagger
> Would that also work?

@marla-singer tried the NPM  package, as we agree it would be ideal. There were [console errors that prevented us from using the NPM package](https://github.com/apinf/platform/issues/2349#issuecomment-291203519), so we settled on the current solution.

**Comment by brylie** [comment]
> Should we open a new ticket so we can review this at a later stage?

Yes please. Using the SwaggerUI NPM package would be ideal. Would you mind opening the follow-up issue?

**Comment by bajiat** [comment]
@mauriciovieira This issue has been waiting for a long time. There is an enhancement issue to try the NPM package. Can we merge this issue as it is, since @marla-singer was not able to use the NPM package?

**Comment by marla-singer** [comment]
@mauriciovieira  @brylie @bajiat 
I found a real reason why it doesn't work properly. Try to fix it today

**Comment by marla-singer** [comment]
@brylie @mauriciovieira Please try again

**Pull request merged**